telemetry/telemetry.go

```diff
@@ -0,0 +1,162 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package telemetry
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"time"
+
+	"github.com/pingcap/errors"
+	"github.com/pingcap/tidb/config"
+	"github.com/pingcap/tidb/sessionctx"
+	"github.com/pingcap/tidb/util/sqlexec"
+	"go.etcd.io/etcd/clientv3"
+)
+
+const (
+	// OwnerKey is the telemetry owner path that is saved to etcd.
+	OwnerKey = "/tidb/telemetry/owner"
+	// Prompt is the prompt for telemetry owner manager.
+	Prompt = "telemetry"
+	// ReportInterval is the interval of the report.
+	ReportInterval = 24 * time.Hour
+)
+
+const (
+	etcdOpTimeout = 3 * time.Second
+	uploadTimeout = 60 * time.Second
+	apiEndpoint   = "https://telemetry.pingcap.com/api/v1/tidb/report"
+)
+
+func getTelemetryGlobalVariable(ctx sessionctx.Context) (bool, error) {
+	rows, _, err := ctx.(sqlexec.RestrictedSQLExecutor).ExecRestrictedSQL(`SELECT @@global.tidb_enable_telemetry`)
+	if err != nil {
+		return false, errors.Trace(err)
+	}
+	if len(rows) != 1 || rows[0].Len() == 0 {
+		return false, fmt.Errorf("unexpected telemetry global variable")
+	}
+	return rows[0].GetString(0) == "1", nil
+}
+
+func isTelemetryEnabled(ctx sessionctx.Context) (bool, error) {
+	if !config.GetGlobalConfig().EnableTelemetry {
+		return false, nil
+	}
+	enabled, err := getTelemetryGlobalVariable(ctx)
+	if err != nil {
+		return false, errors.Trace(err)
+	}
+	return enabled, nil
+}
+
+// PreviewUsageData returns a preview of the usage data that is going to be reported.
+func PreviewUsageData(ctx sessionctx.Context, etcdClient *clientv3.Client) (string, error) {
+	if etcdClient == nil {
+		return "", nil
+	}
+	if enabled, err := isTelemetryEnabled(ctx); err != nil || !enabled {
+		return "", err
+	}
+
+	trackingID, err := GetTrackingID(etcdClient)
+	if err != nil {
+		return "", errors.Trace(err)
+	}
+
+	// note: trackingID may be empty. However, as a preview data, it is fine.
+	data := generateTelemetryData(ctx, trackingID)
+
+	prettyJSON, err := json.MarshalIndent(data, "", "  ")
+	if err != nil {
+		return "", errors.Trace(err)
+	}
+
+	return string(prettyJSON), nil
+}
+
+func reportUsageData(ctx sessionctx.Context, etcdClient *clientv3.Client) (bool, error) {
+	if etcdClient == nil {
+		// silently ignore
+		return false, nil
+	}
+	enabled, err := isTelemetryEnabled(ctx)
+	if err != nil {
+		return false, err
+	}
+	if !enabled {
+		return false, errors.Errorf("telemetry is disabled")
+	}
+
+	trackingID, err := GetTrackingID(etcdClient)
+	if err != nil {
+		return false, errors.Trace(err)
+	}
+	if len(trackingID) == 0 {
+		trackingID, err = ResetTrackingID(etcdClient)
+		if err != nil {
+			return false, errors.Trace(err)
+		}
+	}
+
+	data := generateTelemetryData(ctx, trackingID)
+
+	rawJSON, err := json.Marshal(data)
+	if err != nil {
+		return false, errors.Trace(err)
+	}
+
+	reqCtx, cancel := context.WithTimeout(context.Background(), uploadTimeout)
```

Comment:
`uploadTimeout` is 60s. Do we need to pass a valid `ctx` to this function? If we close the domain, it may get stuck here, which leads to the slow exit of TiDB.